In [1]:
import pandas as pd 
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import datetime as dt

pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_excel('online_retail_II.xlsx')

In [4]:
data.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom

In [6]:
data.tail()

Invoice StockCode                         Description  Quantity  \
525456  538171     22271                FELTCRAFT DOLL ROSIE         2   
525457  538171     22750        FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751      FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970  PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931              JUMBO STORAGE BAG SUKI         2   

               InvoiceDate  Price  Customer ID         Country  
525456 2010-12-09 20:01:00   2.95      17530.0  United Kingdom  
525457 2010-12-09 20:01:00   3.75      17530.0  United Kingdom  
525458 2010-12-09 20:01:00   3.75      17530.0  United Kingdom  
525459 2010-12-09 20:01:00   3.75      17530.0  United Kingdom  
525460 2010-12-09 20:01:00   1.95      17530.0  United Kingdom

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB


In [8]:
data['Country'].nunique()

40

In [9]:
data['Country'].value_counts()

United Kingdom          485852
EIRE                      9670
Germany                   8129
France                    5772
Netherlands               2769
Spain                     1278
Switzerland               1187
Portugal                  1101
Belgium                   1054
Channel Islands            906
Sweden                     902
Italy                      731
Australia                  654
Cyprus                     554
Austria                    537
Greece                     517
United Arab Emirates       432
Denmark                    428
Norway                     369
Finland                    354
Unspecified                310
USA                        244
Japan                      224
Poland                     194
Malta                      172
Lithuania                  154
Singapore                  117
RSA                        111
Bahrain                    107
Canada                      77
Hong Kong                   76
Thailand                    76
Israel  

In [7]:
data.sort_values(by='Price',ascending=False).head()

Invoice     StockCode   Description  Quantity         InvoiceDate  \
241824  C512770             M        Manual        -1 2010-06-17 16:52:00   
241827   512771             M        Manual         1 2010-06-17 16:53:00   
320581  C520667  BANK CHARGES  Bank Charges        -1 2010-08-27 13:42:00   
517953  C537630     AMAZONFEE    AMAZON FEE        -1 2010-12-07 15:04:00   
519294  C537651     AMAZONFEE    AMAZON FEE        -1 2010-12-07 15:49:00   

           Price  Customer ID         Country  
241824  25111.09      17399.0  United Kingdom  
241827  25111.09          NaN  United Kingdom  
320581  18910.69          NaN  United Kingdom  
517953  13541.33          NaN  United Kingdom  
519294  13541.33          NaN  United Kingdom

In [8]:
data.groupby("Description").agg({"Quantity":lambda x : x.sum()}).sort_values('Quantity', ascending=False).head(10)

Quantity
Description                                 
WHITE HANGING HEART T-LIGHT HOLDER     57733
WORLD WAR 2 GLIDERS ASSTD DESIGNS      54698
BROCADE RING PURSE                     47647
PACK OF 72 RETRO SPOT CAKE CASES       46106
ASSORTED COLOUR BIRD ORNAMENT          44925
60 TEATIME FAIRY CAKE CASES            36326
PACK OF 60 PINK PAISLEY CAKE CASES     31822
JUMBO BAG RED RETROSPOT                30727
SMALL POPCORN HOLDER                   29500
STRAWBERRY CERAMIC TRINKET BOX         26563

In [9]:
print(f"Number of uniques in StockCode: {data['StockCode'].nunique()}")
print(f"Number of uniques in Description: {data['StockCode'].nunique()}")

Number of uniques in StockCode: 4632
Number of uniques in Description: 4632


In [10]:
data.dropna(inplace=True)

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417534 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      417534 non-null  object        
 1   StockCode    417534 non-null  object        
 2   Description  417534 non-null  object        
 3   Quantity     417534 non-null  int64         
 4   InvoiceDate  417534 non-null  datetime64[ns]
 5   Price        417534 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      417534 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 28.7+ MB


In [12]:
data.describe()

Quantity          Price    Customer ID
count  417534.000000  417534.000000  417534.000000
mean       12.758815       3.887547   15360.645478
std       101.220424      71.131797    1680.811316
min     -9360.000000       0.000000   12346.000000
25%         2.000000       1.250000   13983.000000
50%         4.000000       1.950000   15311.000000
75%        12.000000       3.750000   16799.000000
max     19152.000000   25111.090000   18287.000000

In [13]:
data.describe([0.01, 0.05, 0.10, 0.25, 0.50, 0.75, 0.90, 0.95, 0.99]).T

count          mean          std      min        1%        5%  \
Quantity     417534.0     12.758815   101.220424  -9360.0     -2.00      1.00   
Price        417534.0      3.887547    71.131797      0.0      0.29      0.42   
Customer ID  417534.0  15360.645478  1680.811316  12346.0  12435.00  12725.00   

                  10%       25%       50%       75%       90%      95%  \
Quantity         1.00      2.00      4.00     12.00     24.00     36.0   
Price            0.65      1.25      1.95      3.75      6.75      8.5   
Customer ID  13042.00  13983.00  15311.00  16799.00  17706.00  17913.0   

                  99%       max  
Quantity       144.00  19152.00  
Price           14.95  25111.09  
Customer ID  18196.00  18287.00

In [14]:
data = data[~data["Invoice"].str.contains("C", na=False)]

In [15]:
data.describe([0.01, 0.05, 0.10, 0.25, 0.50, 0.75, 0.90, 0.95, 0.99]).T

count          mean          std      min        1%        5%  \
Quantity     407695.0     13.586686    96.842229      1.0      1.00      1.00   
Price        407695.0      3.294188    34.756655      0.0      0.29      0.42   
Customer ID  407695.0  15368.504107  1679.795700  12346.0  12435.00  12731.00   

                  10%       25%       50%       75%       90%      95%  \
Quantity         1.00      2.00      5.00     12.00     24.00     36.0   
Price            0.65      1.25      1.95      3.75      6.75      8.5   
Customer ID  13044.00  13997.00  15321.00  16812.00  17706.00  17913.0   

                  99%      max  
Quantity       144.00  19152.0  
Price           14.95  10953.5  
Customer ID  18196.00  18287.0

In [16]:
data["TotalPrice"] = data["Quantity"] * data["Price"]

In [17]:
data.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom        83.4  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom        81.0  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom        81.0  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom       100.8  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom        30.0

RFM Metrics

# The last date of purchase

In [18]:
data['InvoiceDate'].max()

Timestamp('2010-12-09 20:01:00')

assign today's date as 2 days after the last date of purchase to make sure that no of the recency values becomes zero

In [19]:
today_date = data["InvoiceDate"].max()+ dt.timedelta(days=2)
today_date

Timestamp('2010-12-11 20:01:00')

create a new data called rfm in order to calculate Recency, Frequency and Monetary values
data is grouped by customers and:
    
1. The number of days between today_date and the last purchase date of the customer in Recency
2. The number of unique invoice of this customer is Frequency 
3. The sum of TotalPrice is this customer's Monetary

In [20]:
rfm = data.groupby('Customer ID').agg({'InvoiceDate': lambda date: (today_date - date.max()).days,
                                       'Invoice': lambda inv: inv.nunique(),
                                       'TotalPrice': lambda price: price.sum()})

In [21]:
rfm.head()

InvoiceDate  Invoice  TotalPrice
Customer ID                                  
12346.0              166       11      372.86
12347.0                4        2     1323.32
12348.0               75        1      222.16
12349.0               44        3     2671.14
12351.0               12        1      300.93

In [22]:
rfm.columns = ["Recency","Frequency", 'Monetary']
rfm.head()

Recency  Frequency  Monetary
Customer ID                              
12346.0          166         11    372.86
12347.0            4          2   1323.32
12348.0           75          1    222.16
12349.0           44          3   2671.14
12351.0           12          1    300.93

In [23]:
rfm.describe([0.01, 0.05, 0.10, 0.25, 0.50, 0.75, 0.90, 0.95, 0.99]).T

count         mean          std  min       1%        5%      10%  \
Recency    4314.0    92.269124    96.943482  2.0   2.0000    4.0000    9.000   
Frequency  4314.0     4.454103     8.168658  1.0   1.0000    1.0000    1.000   
Monetary   4314.0  2047.288659  8912.523243  0.0  39.9565  110.4365  155.485   

              25%     50%        75%       90%        95%         99%  \
Recency     19.00   54.00   137.0000   256.000   305.0000    369.0000   
Frequency    1.00    2.00     5.0000     9.000    13.0000     31.0000   
Monetary   307.95  705.55  1722.8025  3796.595  6235.1205  20136.7825   

                 max  
Recency       375.00  
Frequency     205.00  
Monetary   349164.35

# RFM Scores

1. The min number of Recency metric means that this customer has just purchased,
   the highest score(5) should be given to the lower number of Recency.
2. The max number of Frequency and Monetary metrics mean that the customer is 
    purchasing Frequently and spending more money, so the highest score (5)  should be
    given to the highest Frequency and Monetary values. 

In [24]:
rfm["RecencyScore"] = pd.qcut(rfm['Recency'], 5, labels=[5, 4, 3, 2, 1])

rfm["FrequencyScore"] = pd.qcut(rfm['Frequency']. rank(method="first"),5, labels=[1,2,3,4,5])

rfm["MonetaryScore"] = pd.qcut(rfm['Monetary'], 5, labels = [1, 2, 3, 4, 5])

In [25]:
rfm["RFM_SCORE"] = (rfm['RecencyScore'].astype(str) +
                    rfm['FrequencyScore'].astype(str) +
                    rfm['MonetaryScore'].astype(str))

rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346.0          166         11    372.86            2              5   
12347.0            4          2   1323.32            5              2   
12348.0           75          1    222.16            2              1   
12349.0           44          3   2671.14            3              3   
12351.0           12          1    300.93            5              1   

            MonetaryScore RFM_SCORE  
Customer ID                          
12346.0                 2       252  
12347.0                 4       524  
12348.0                 1       211  
12349.0                 5       335  
12351.0                 2       512

Display some of the customers with the highest Scores:

In [26]:
rfm[rfm['RFM_SCORE'] == "555"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12415.0           12          7  19543.84            5              5   
12431.0           10         13   4370.52            5              5   
12471.0           11         49  20139.74            5              5   
12472.0            6         13  11308.48            5              5   
12474.0           15         13   5048.66            5              5   

            MonetaryScore RFM_SCORE  
Customer ID                          
12415.0                 5       555  
12431.0                 5       555  
12471.0                 5       555  
12472.0                 5       555  
12474.0                 5       555

# Naming The RFM Scores 

The following dict has been made according tothe famous RFM graphic

In [27]:
seg_map = {
    r'[1-2][1-2]': 'Hibernating',
    r'[1-2][3-4]': 'At_Risk',
    r'[1-2]5': 'Cant_Loose',
    r'3[1-2]': 'About_to_Sleep',
    r'33': 'Need_Attention',
    r'[3-4][4-5]': 'Loyal_Customers',
    r'41': 'Promising',
    r'51': 'New_Customers',
    r'[4-5][2-3]': 'Potential_Loyalists',
    r'5[4-5]': 'Champions'
}

We will be using Recency and Frequency scores for customer segmentation. We are
assuming that a customer who has recently purchased and who is often purchasing should
have high RFM scores.

In [28]:
rfm['Segment'] = rfm['RecencyScore'].astype(str) + rfm['FrequencyScore'].astype(str)
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346.0          166         11    372.86            2              5   
12347.0            4          2   1323.32            5              2   
12348.0           75          1    222.16            2              1   
12349.0           44          3   2671.14            3              3   
12351.0           12          1    300.93            5              1   

            MonetaryScore RFM_SCORE Segment  
Customer ID                                  
12346.0                 2       252      25  
12347.0                 4       524      52  
12348.0                 1       211      21  
12349.0                 5       335      33  
12351.0                 2       512      51

In this step We will convert the metrics into category names

In [29]:
rfm['Segment'] = rfm['Segment'].replace(seg_map, regex=True)
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346.0          166         11    372.86            2              5   
12347.0            4          2   1323.32            5              2   
12348.0           75          1    222.16            2              1   
12349.0           44          3   2671.14            3              3   
12351.0           12          1    300.93            5              1   

            MonetaryScore RFM_SCORE              Segment  
Customer ID                                               
12346.0                 2       252           Cant_Loose  
12347.0                 4       524  Potential_Loyalists  
12348.0                 1       211          Hibernating  
12349.0                 5       335       Need_Attention  
12351.0                 2       512        New_Customers

See the number of customers that fall into each category

In [30]:
rfm['Segment'].value_counts()

Hibernating            1017
Loyal_Customers         742
Champions               663
At_Risk                 611
Potential_Loyalists     517
About_to_Sleep          343
Need_Attention          207
Promising                87
Cant_Loose               77
New_Customers            50
Name: Segment, dtype: int64

Now that we have all the scores for the customers and we have been able to categorize them into 10 groups, 
it's time for action. We will be using metrics for this process, not scores. 
We will be focusing on the groups that need a better customer relationship and try to figure out 
what we can do in order to make that specific segment purchase more frequently and become loyal. 
Thanks to RFM scores of the segments, we know what exactly that segment needs.

In [31]:
rfm[["Segment", "Recency", "Frequency", "Monetary"]].groupby("Segment").agg(["mean", "count"])

Recency        Frequency           Monetary      
                           mean count       mean count         mean count
Segment                                                                  
About_to_Sleep        54.819242   343   1.201166   343   441.320000   343
At_Risk              153.155483   611   3.073650   611  1188.878316   611
Cant_Loose           125.116883    77   9.116883    77  4099.450000    77
Champions              8.119155   663  12.553544   663  6852.264167   663
Hibernating          215.056047  1017   1.125860  1017   403.183386  1017
Loyal_Customers       37.287062   742   6.830189   742  2746.067353   742
Need_Attention        54.265700   207   2.449275   207  1060.357005   207
New_Customers          9.580000    50   1.000000    50   386.199200    50
Potential_Loyalists   19.793037   517   2.017408   517   729.510986   517
Promising             26.747126    87   1.000000    87   367.086782    87

the customer that need attention

In [32]:
rfm[rfm["Segment"] == "Need_Attention"]

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12349.0           44          3   2671.14            3              3   
12369.0           50          3   1791.15            3              3   
12371.0           46          3   2179.42            3              3   
12374.0           58          3   2246.29            3              3   
12389.0           39          3   1433.33            3              3   
...              ...        ...       ...          ...            ...   
18078.0           62          2    508.47            3              3   
18112.0           55          2    423.27            3              3   
18136.0           46          2    847.03            3              3   
18141.0           38          2    882.47            3              3   
18244.0           67          2    568.64            3              3   

            MonetaryScore RFM_SCORE         Segment  
Customer ID                                          
12349.0                 5       335  Need_Attention  
12369.0                 4       334  Need_Attention  
12371.0                 5       335  Need_Attention  
12374.0                 5       335  Need_Attention  
12389.0                 4       334  Need_Attention  
...                   ...       ...             ...  
18078.0                 3       333  Need_Attention  
18112.0                 2       332  Need_Attention  
18136.0                 3       333  Need_Attention  
18141.0                 3       333  Need_Attention  
18244.0                 3       333  Need_Attention  

[207 rows x 8 columns]

Need_Attention segment has 207 customers that last purchased nearly 2 months ago, 
despite the fact that they don't frequently purchase, they spend quite a good amount of money. 
So we should be focusing on this group. In order to make them transform into a customer that purchases frequently, 
we can offer them some discount with a time limit of 30 days. 
So that they would revisit and purchase.

In [33]:
marketing_df = pd.DataFrame()
marketing_df['Cant_Losse'] = rfm[rfm["Segment"]=="Cant_Loose"].index

Can't_Loose segment has purchased for 9 times this year but the last date of this was nearly 4 months, they spend a good amount of money and they used to be our loyal customers, we can't loose them. We should put this 77 customers into our loyalty program, offer them seasonal discounts, make them feel special while purchasing from our company and make them loyal again. We can export the customer id list into an excel file and pass this file to our Marketing Department.

change the dtype of Customer ID variable in order to get rid of the decimal par

In [40]:
marketing_df['Cant_Losse']=marketing_df['Cant_Losse'].astype(int)

In [43]:
marketing_df.info

<bound method DataFrame.info of     Cant_Losse
0        12346
1        12380
2        12482
3        12510
4        12891
..         ...
72       18051
73       18064
74       18094
75       18251
76       18258

[77 rows x 1 columns]>

In [44]:
marketing_df.to_csv("cant_loose.csv")